cmake/version.cmake

```diff
@@ -61,8 +61,12 @@ endif()
 
 set(PROJECT_VERSION_STR ${PROJECT_VERSION}${PROJECT_VERSION_EXTRA_STR})
 
+if(DEFINED BUILD_VERSION)
+  set(BUILD_VERSION_STR ", build: ${BUILD_VERSION}")
```

Comment:
Why not keep `BUILD_VERSION:`? It's searchable in both the logs and source code whereas `build:` is not.

Reply:
I'm with @marc-hb on this one

Reply:
I disagree, the messages written to the user during CMake time should be for the user and not CI.
Of course if CI can use the info if they find that useful, I see no problems there, but we shouldn't format our messages because of CI.

Basing log messages on downstream CIs would put unnecessary restrictions on ourselves and limit our freedom in changing the message printing.

Also today, we print:
```
Including boilerplate (Zephyr repository): /projects/github/ncs/zephyr/cmake/app/boilerplate.cmake
```
not `ZEPHYR_BASE=/projects/github/ncs/zephyr`.

```
-- Zephyr version: 2.5.99 (/projects/github/ncs/zephyr)
```
not `PROJECT_VERSION=2.5.99`

```
-- Board: nrf52840dk_nrf52840
```
not `BOARD=nrf52840dk_nrf52840`

In fact, we don't print any CMake variables today.

If we should make this nice for CI integration (which I think is a valuable suggestion in itself), then I would prefer we place the `BUILD_VERSION` in the `CMakeCache.txt` as an internal cache variable when not provided by the user.

That will provide a guaranteed way for CI to always lookup the BUILD_VERSION value and not rely on print message output, while at the same time allow us to format the messages as we find most appropriate for users.

Reply:
I was maybe a bit fast on the CI reference, which comes from the original reason for adding build version to log output here: https://github.com/zephyrproject-rtos/zephyr/pull/34842#issue-630159511

but which isn't directly mentioned in @marc-hb comment.

But my reasoning for not using `BUILD_VERSION=<value>` still stands.

Reply:
> I disagree, the messages written to the user during CMake time should be for the user and not CI.

I didn't have CI in mind at all. When I wrote "searchable in the logs and in the source" I only had users in mind.

Reply:
I think @tejlmand has a fair point.

Reply:
> A `build:` prefix is unsearchable and useless.

@marc-hb What's wrong with this ?
```
$ grep 'Zephyr version' build.log 
-- Zephyr version: 2.5.99 (/projects/github/ncs/zephyr), build: zephyr-v2.5.0-3541-g3cce8527ee18
```
That seems like a quite unique search pattern in a build log, and give you all relevant build version in a single line.
The Zephyr version itself, the path to the Zephyr, and the `git describe` / user specified BUILD_VERSION.

> Also, why do you think -Werror exists?

Not sure I see the relevance.
CMake already print warnings and errors, including reference to where to message originated.
Are you asking for the possibility for the user to stop CMake processing in case of warnings ? (errors already stops CMake processing)

> Not printing CMake variables is generally a mistake 

No, but people should understand the difference between log levels.
When I run `ninja,` I get warnings and errors printed and important messages.

I don't get all compile invocations printed. If I want to know everything then I use `ninja -v`.
But you often see people failing to properly specify the log message, and just print everything per default, just like OpenThread, where running CMake resullts in 40 lines of variables like this:
```
OPENTHREAD_CONFIG_ASSERT_ENABLE=1
OPENTHREAD_CONFIG_BACKBONE_ROUTER_DUA_NDPROXYING_ENABLE=0
OPENTHREAD_CONFIG_BACKBONE_ROUTER_MULTICAST_ROUTING_ENABLE=0
OPENTHREAD_CONFIG_BORDER_AGENT_ENABLE=1
OPENTHREAD_CONFIG_BORDER_ROUTER_ENABLE=1
....
```
Since CMake 3.15, it has been possible to use `message(VERBOSE "<VAR>=<VAL>")` so that they are only printed when using: `cmake --log-level=VERBOSE`, so that users decides, just like using `ninja` or `ninja -v`.

Feel free to propose a PR that will print Zephyr CMake variables when using `cmake --log-level=VERBOSE`.
I think that could be very useful.
Just don't print the variables per default, but allow the user to decide when those should be printed with `--log-level=VERBOSE`.

Reply:
> What's wrong with this ? `$ grep 'Zephyr version' build.log` 

You have to know the "magic words". Before this PR you could do:
1. `$ cd zephyr; git grep 'git describe' # -> points at git.cmake`
2. Open git.cmake, find BUILD_VERSION
3. Search for BUILD_VERSION anywhere, done.

You added extra and unnecessary steps. Also, searching for BUILD_VERSION has no risk of whitespace of case mistake.

> > Also, why do you think -Werror exists?

> Not sure I see the relevance.

-Werror was invented because people NEVER read logs and never see warnings. They only skim or search logs (BTW neither Github, Travis or Jenkins has any concept of a warning: it's either red or green, nothing in between)

> Feel free to propose a PR that will print Zephyr CMake variables when using cmake --log-level=VERBOSE

That's really not my point. Everyone agreed to print BUILD_VERSION at the default log level, my point is just: why give it a different name? This adds indirection and some confusion for zero value.

(and yes log levels are of course useful, different topic again)